### PR TITLE
Improvements in `picky-asn1-x509` and `picky-krb` related to PKU2U

### PR DESCRIPTION
**Pull request by RRRadicalEdward**
I've added new features needed for the PKU2U protocol.

List of the improvements:

* `picky-asn1-x509`:
    * Implemented `DirectoryString::BmpString` because Pku2u certificated can contain BmpString. It had been commented out. 
    * I've made macros exportable because I need them to deserialize new types from `picky-krb`.
    * Added more OIDs.
* `picky-krb`:
    * Added `negoex` module that contains all new structures and data types needed for the NEGOEX protocol.
    * Added `pkinit` module with implemented structures and data type needed for the PKU2U protocol.
    * Added a few new structures from `MS-KILE` needed for the final authenticator generation (`KerbAdRestrictionEntry`, `LsapTokenInfoIntegrity`).
    * Added new needed constants: key usages, GSS API token IDs, and checksum types.
    * The `crypto` module:
        * Implemented Diffie-Hellman key exchange algorithm that is used in Pku2u.
        * Added `Checksum` trait and implement it for the `HmacSha196Aes128/256` and `HmacSha1Des3Kd`.
        * Added the `Cipher::random_to_key` method that allows us to generate an encryption key from any input vector. For example, from DH shared secret. Not only from the password as it was before. Implement this method for all supported encryption types.

Used Docs & References:

* [PKINIT](https://www.rfc-editor.org/rfc/rfc4556)
* [NEGOEX](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-negoex/77c795cf-e522-4678-b0f1-2063c5c0561c)
* [PKU2U draft](https://datatracker.ietf.org/doc/html/draft-zhu-pku2u-09)
* [MS-KILE](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-kile/2a32282e-dd48-4ad9-a542-609804b02cc9)
* [Encryption and Checksum Specifications for Kerberos 5](https://www.rfc-editor.org/rfc/rfc3961.html)
* [Diffie-Hellman Key Agreement Method](https://www.rfc-editor.org/rfc/rfc2631)

### CODE REVIEW

**Comment by RRRadicalEdward** [comment]
Let me make CI happy before reviewing  🙂

**Comment by RRRadicalEdward** [comment]
@CBenoit This PR is ready for reviewing 🙂

**Comment by CBenoit** [comment]
> @CBenoit This PR is ready for reviewing 🙂

Sorry I’ve been busy. I’ll review ASAP

File: picky-krb/Cargo.toml
Comment (CBenoit) [inline picky-krb/Cargo.toml:38]: What is motivating `bincode` addition? I don’t see any usage 🤔

File: picky-krb/src/crypto/aes/key_derivation.rs
Comment (CBenoit) [inline picky-krb/src/crypto/aes/key_derivation.rs:18]: The input is returned as-is. Is that intended?

**Review by CBenoit — COMMENTED** [review]
I still have a few more files to review, but here is my initial feedback 🙂

File: picky-asn1-x509/src/macros.rs
Comment (RRRadicalEdward) [reply picky-asn1-x509/src/macros.rs:1]: I use those macros for the PaPkAsRep deserialization. I made them private again and rewrote the deserialization without macros.

File: picky-asn1/src/bit_string.rs
Comment (RRRadicalEdward) [reply picky-asn1/src/bit_string.rs:45]: Yes, you are right. I have written it as a draft method and forgot to change it

File: picky-krb/Cargo.toml
Comment (RRRadicalEdward) [reply picky-krb/Cargo.toml:38]: I've planned to use it for the NEGOEX data types parsing/encoding but it turned out not to be convenient. I will remove it

File: picky-krb/src/crypto/aes/key_derivation.rs
Comment (RRRadicalEdward) [reply picky-krb/src/crypto/aes/key_derivation.rs:18]: Yes, it is intended. I made this function more for code readability than for functionality. According to RFC 3961, we have the rundom_to_key function. In AES encryption it does nothing(`irandom-to-key function: identity function` https://www.rfc-editor.org/rfc/rfc3962.html#section-6)

File: picky-asn1-x509/src/lib.rs
Comment (RRRadicalEdward) [reply picky-asn1-x509/src/lib.rs]: Fixes

File: picky-asn1-x509/src/directory_string.rs
Comment (RRRadicalEdward) [reply picky-asn1-x509/src/directory_string.rs]: Done

File: picky-krb/Cargo.toml
Comment (RRRadicalEdward) [reply picky-krb/Cargo.toml]: I replaced `num-bigint` by `num-bigint-dig`

File: picky-asn1-x509/src/macros.rs
Comment (RRRadicalEdward) [reply picky-asn1-x509/src/macros.rs]: I removed the `#[macro_export]`s

File: picky-asn1-der/Cargo.toml
Comment (RRRadicalEdward) [reply picky-asn1-der/Cargo.toml:20]: I added the version

File: picky-krb/Cargo.toml
Comment (RRRadicalEdward) [reply picky-krb/Cargo.toml]: I added the versions

File: picky-krb/src/crypto/aes/key_derivation.rs
Comment (CBenoit) [reply picky-krb/src/crypto/aes/key_derivation.rs:18]: Makes sense! I think it would help to have this as a comment in the code 🙂

File: Cargo.lock
Comment (CBenoit) [reply Cargo.lock]: I see this is an optional dependency. Did you try to add `default-features = false` to the `chrono` dependency we have in `picky`?

File: picky-krb/src/negoex/data_types.rs
Comment (CBenoit) [inline picky-krb/src/negoex/data_types.rs:71]: Nitpick. I feel it’s a bit overkill to import and use a proc-macro crate (`num-derive`) for a single struct. Not sure it’s significant in the current state though. Not a blocker.

**Review by CBenoit — COMMENTED** [review]
Looking good! We can merge once last concerns are addressed! 😄

File: picky-krb/src/crypto/aes/key_derivation.rs
Comment (RRRadicalEdward) [reply picky-krb/src/crypto/aes/key_derivation.rs:24]: I added a comment

File: Cargo.lock
Comment (CBenoit) [reply Cargo.lock]: Thank you, no need to dig further then

File: picky-krb/src/negoex/data_types.rs
Comment (RRRadicalEdward) [reply picky-krb/src/negoex/data_types.rs:71]: Don't merge for now. Let me fix it. I'm working on it

Comment (RRRadicalEdward) [reply picky-krb/src/negoex/data_types.rs]: Just pushed a commit with the fix

File: picky-krb/Cargo.toml
Comment (CBenoit) [inline picky-krb/Cargo.toml:35]: Is `num-traits` still used?

**Comment by RRRadicalEdward** [comment]
@CBenoit CI started to fail on the `Lints WASM` job somewhy. Is it related to my code?

**Review by CBenoit — APPROVED** [review]
> CI started to fail on the `Lints WASM` job somewhy. Is it related to my code?

It doesn’t sound related 🤔 
I’ll have a look at what’s happening

In any event, the PR looks good to me 🙂

**Comment by CBenoit** [comment]
I believe https://github.com/Devolutions/picky-rs/pull/187 might fix it, you’ll need to rebase on the top once it’s merged 🙂

**Comment by CBenoit** [comment]
I went ahead and rebased your branch on top of master.

**Pull request merged**
